scheduler/src/cook/kubernetes/api.clj

```diff
@@ -14,32 +14,59 @@
     (java.util.concurrent Executors ExecutorService)))
 
 
+(def cook-pod-label "twosigma.com/cook-scheduler-job")
+
 (def ^ExecutorService kubernetes-executor (Executors/newFixedThreadPool 2))
 
+(defn is-cook-scheduler-pod
+  "Is this a cook pod? Uses some-> so is null-safe."
+  [^V1Pod pod]
+  (some-> pod .getMetadata .getLabels (.get cook-pod-label)))
+
+(defn make-atom-updater
+  "Given a state atom, returns a callback that updates that state-atom when called with a key, prev item, and item."
+  [state-atom]
+  (fn
+    [key prev-item item]
+    (cond
+      (and (nil? prev-item) (not (nil? item))) (swap! state-atom (fn [m] (assoc m key item)))
+      (and (not (nil? prev-item)) (not (nil? item))) (swap! state-atom (fn [m] (assoc m key item)))
+      (and (not (nil? prev-item)) (nil? item)) (swap! state-atom (fn [m] (dissoc m key))))))
+
+(defn cook-pod-callback-wrap
+  "A special wrapping function that, given a callback, key, prev-item, and item, will invoke the callback only
+  if the item is a pod that is a cook scheduler pod. (THe idea is that (partial cook-pod-callback-wrap othercallback)
+  returns a new callback that only invokes othercallback on cook pods."
+  [callback key prev-item item]
+  (when (or (is-cook-scheduler-pod prev-item) (is-cook-scheduler-pod item))
+    (callback key prev-item item)))
+
 (defn handle-watch-updates
   "When a watch update occurs (for pods or nodes) update both the state atom as well as
-  invoke the callback on the previous and new values for the key."
-  [state-atom ^Watch watch key-fn callback]
+  invoke the callbacks on the previous and new values for the key."
+  [state-atom ^Watch watch key-fn callbacks]
   (while (.hasNext watch)
-    (let [update (.next watch)
-          item (.-object update)
-          prev-item (get @state-atom (key-fn item))]
-      (case (.-type update)
-        "ADDED" (swap! state-atom (fn [m] (assoc m (key-fn item) item)))
-        "MODIFIED" (swap! state-atom (fn [m] (assoc m (key-fn item) item)))
-        "DELETED" (swap! state-atom (fn [m] (dissoc m (key-fn item)))))
-      (when callback
+    (let [watch-response (.next watch)
+          item (.-object watch-response)
+          key (key-fn item)
+          prev-item (get @state-atom key)]
+      (doseq [callback callbacks]
         (try
-          (callback prev-item item)
+          (callback key prev-item item)
```

Comment:
What do you think about just passing `state-atom` and `item` to the callback? Then you can drop the `key-fn` here.

Reply:
It makes this API less confusing and harder to misuse. 
1. There's more than one state atom, so which one gets passed here (and why). E.g., the callback for updating the cook pods atom would also be passed the all-pods atom?
2. The callback that forwards to cook doesn't use the state atom.
3. We move the key lookup into each of the callee's of this function. So we now have mulitple locations that are using it now.